circe-akka-serializer/src/main/scala/org/virtuslab/ash/circe/Compression.scala

```diff
@@ -0,0 +1,60 @@
+package org.virtuslab.ash.circe
+
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.util.zip.GZIPInputStream
+import java.util.zip.GZIPOutputStream
+
+import scala.annotation.tailrec
+
+object Compression {
+  sealed trait Algorithm
+  case object Off extends Algorithm
+  case class GZip(greaterThan: Long) extends Algorithm
+  // TODO (#159): add support for LZ4 java compression
+  // case class LZ4(greaterThan: Long) extends Algorithm
+
+  private[circe] def compressIfNeeded(
+      bytes: Array[Byte],
+      bufferSize: Int,
+      compressionAlgorithm: Algorithm): Array[Byte] =
+    compressionAlgorithm match {
+      case Compression.Off => bytes
+      case Compression.GZip(largerThan) =>
+        if (bytes.length <= largerThan) {
+          bytes
+        } else {
+          val byteArrayOutputStream = new ByteArrayOutputStream(bufferSize)
+          val outputStream = new GZIPOutputStream(byteArrayOutputStream)
+          try outputStream.write(bytes)
+          finally outputStream.close()
+          byteArrayOutputStream.toByteArray
+        }
+    }
+
+  private[circe] def decompressIfNeeded(bytes: Array[Byte], bufferSize: Int): Array[Byte] =
+    if (isCompressedWithGzip(bytes)) {
+      val inputStream = new GZIPInputStream(new ByteArrayInputStream(bytes))
+      val outputStream = new ByteArrayOutputStream()
+      val buffer = new Array[Byte](bufferSize)
+
+      @tailrec def readChunk(): Unit =
+        inputStream.read(buffer) match {
+          case -1 => ()
+          case n =>
+            outputStream.write(buffer, 0, n)
+            readChunk()
+        }
+
+      try readChunk()
+      finally inputStream.close()
+      outputStream.toByteArray
+    } else {
+      bytes
+    }
+
+  private[circe] def isCompressedWithGzip(bytes: Array[Byte]): Boolean =
+    (bytes != null) && (bytes.length >= 2) &&
+    (bytes(0) == GZIPInputStream.GZIP_MAGIC.toByte) &&
```

Comment:
Is it theoretically possible that a payload which is NOT compressed with gzip will still start with these 2 bytes? 🤔

Reply:
(and hence incorrectly recognized as compressed)

Reply:
I am not 100% sure, but this GZIP_MAGIC constant is a standard for all .gz files and also - such check is done in akka-serialization-jackson as well (which is used as a standard for akka cluster). See: https://github.com/akka/akka/blob/main/akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonSerializer.scala#L80

Reply:
So I think we can assume that this is safe

Reply:
Okay... AFAICS `GZIP_MAGIC` is 35615 aka `8b1f` in Hex... both characters being outside of ASCII range.

Could you confirm that **without compression, all characters in the serialized data are from ASCII range**?
If that assumption is correct, I'd even add a test to demonstrate that.

> such check is done in akka-serialization-jackson as well (which is used as a standard for akka cluster)

I wouldn't trust that even if this is safe for `akka-serialization-jackson`, this will still be safe for our serializer... unless the above holds true ☝🏻

Reply:
Ok, I did some research and debugging with printlns and that's what I've found:
  
1. After my changes there are two ways of converting a serializable object to binary (to Array[Byte]) and it happens inside CirceAkkaSerializer#toBinary method. Interesting part of code in this method is as follows:
```
val bytes = printer.print(encoder.asInstanceOf[Encoder[AnyRef]](o)).getBytes(UTF_8)
val result = Compression.compressIfNeeded(bytes, bufferSize, compressionAlgorithm)
```
  
And compressIfNeeded might do compression or just return raw bytes defined earlier. And we want to be sure, that if serialization is done without compression (raw bytes passed) - then it will not be possible that:
```
isCompressedWithGzip(bytes) == true
```
  
2. So, I started checking what exactly is produced as `bytes` inside CirceAkkaSerializer#toBinary method:
  ```
 val bytes = printer.print(encoder.asInstanceOf[Encoder[AnyRef]](o)).getBytes(UTF_8)
```
`printer.print` "Returns a string representation of a pretty-printed JSON value." (javadoc quoted)
  And afterwards this String is encoded into byte array (`getBytes`) using UTF_8
  
Since we are encoding JSON objects - they will start with '{' char. Thus, the first element of the `bytes` array is always the 123 Byte number - which is the decimal representation of the `{` character (123 decimal is the same as 7b hexadecimal).
  
3. I diged the `GZIP_MAGIC` number - which is a java int with the value of 35615 (8b1f hex). One of the conditions in our test is as below:
```
(bytes(0) == GZIPInputStream.GZIP_MAGIC.toByte)
```
which gets evaluated to:  
```
bytes(0) == 35615.toByte
```
which gets evaluated to:  

```
bytes(0) == 31 // but 31 is of type Byte ...
```
  
And since bytes(0) holds a Byte with value equal to 123 - this will never be true.
  
JFYI - 31 is a decimal representation of the '=' sign
  
So, in my opinion, this will work well and there is no reason to worry. Since the flow goes like this:  
JSON object.   >.   String object.   >.   Array[Byte] object. 
which means that before transforming into byte array, CIRCE takes care of creating Json from our object and Json must not start with '=' character

Reply:
> JFYI - 31 is a decimal representation of the '=' sign

Hmmm 31 (aka `1f` in hex) is actually a non-printable character: `UNIT SEPARATOR (US) DOWN ARROW`: https://www.fileformat.info/info/unicode/char/1f/index.htm

But still, that's even less likely to occur in JSON data, let alone as the first character 😅 

Other than that &mdash; a brilliant analysis, thanks! if you could put an **abridged** version of that as a comment in code, so that the future generation don't scratch their heads

Reply:
> Since we are encoding JSON objects - they will start with '{' char.

For clarity &mdash; we're always encoding JSON **objects**, as in `{...}`, right? can't we ever encode a JSON number/string/array as well?

Reply:
We are encoding objects extending AnyRef type - so it could be either a Json object (starting with '{' character - 123 decimal) or a Json Array (starting with '[' character - 91 decimal) - but this second option would rather be very rare in my opinion. So - for both cases GZIP_MAGIC number will work well.

Reply:
Comment, added - please kindly review. :)